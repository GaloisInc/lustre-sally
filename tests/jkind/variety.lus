-- This file uses a variety of Lustre features
-- Expected output:
-- VALID PROPERTIES: [ok1]
-- INVALID PROPERTIES: [cex1] (k = 10)

type point = struct {x : int; y : int};

const ORIGIN = point {x = 0; y = 0};

node add(p, q : point) returns (r : point);
var
  x_coord, y_coord : int;
let
  x_coord = p.x + q.x;
  y_coord = p.y + q.y;
  r = point {x = x_coord; y = y_coord};
tel;

node abs(z : int) returns (az : int);
let
  az = if z >= 0 then z else -z;
tel;

node main(delta_x, delta_y : int) returns ();
var
  p1, p2, delta : point;
  ok1, cex1 : bool;
let
  p1 = point {x = (0 -> pre p1.x) + delta_x;
              y = (0 -> pre p1.y) + delta_y};

  delta = point {x = delta_x; y = delta_y};
  p2 = add(ORIGIN -> pre p2, delta);

  assert abs(delta_x) <= 2;
  assert abs(delta_y) <= 2;

  ok1 = p1 = p2;
  --%PROPERTY ok1;

  cex1 = p1 <> point {x = 20; y = 13};
  --%PROPERTY cex1;
tel;
