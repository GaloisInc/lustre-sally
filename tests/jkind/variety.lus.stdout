Sally Model
===========

(define-state-type
   S
   (
     (Leq Bool)
     (abs Int)
     (cex1 Bool)
     (ok1 Bool)
     (z Int)
     (|Leq:1| Bool)
     (|abs:1| Int)
     (|delta.x| Int)
     (|delta.y| Int)
     (|delta_x| Int)
     (|delta_y| Int)
     (|gal-initializing| Bool)
     (|init_pre_p1.x:init| Bool)
     (|init_pre_p1.x| Int)
     (|init_pre_p1.y:init| Bool)
     (|init_pre_p1.y| Int)
     (|p.x:init| Bool)
     (|p.x| Int)
     (|p.y:init| Bool)
     (|p.y| Int)
     (|p1.x| Int)
     (|p1.y| Int)
     (|p2.x| Int)
     (|p2.y| Int)
     (|pre_p1.x| Int)
     (|pre_p1.y| Int)
     (|pre_p2.x| Int)
     (|pre_p2.y| Int)
     (|q.x| Int)
     (|q.y| Int)
     (|x_coord| Int)
     (|y_coord| Int)
     (|z:1| Int))
   (
     (|delta_x| Int)
     (|delta_y| Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |init_pre_p1.x:init| true)
      (= |init_pre_p1.y:init| true)
      (= |p.x:init| true)
      (= |p.y:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|delta_x| input.|delta_x|)
      (= next.|delta_y| input.|delta_y|)
      next.Leq
      next.|Leq:1|
      (= next.|pre_p1.x| state.|p1.x|)
      (ite
         state.|init_pre_p1.x:init|
         (= next.|init_pre_p1.x| 0)
         (= next.|init_pre_p1.x| next.|pre_p1.x|))
      (= next.|init_pre_p1.x:init| false)
      (=
         next.|p1.x|
         (+ next.|init_pre_p1.x| next.|delta_x|))
      (= next.|pre_p1.y| state.|p1.y|)
      (ite
         state.|init_pre_p1.y:init|
         (= next.|init_pre_p1.y| 0)
         (= next.|init_pre_p1.y| next.|pre_p1.y|))
      (= next.|init_pre_p1.y:init| false)
      (=
         next.|p1.y|
         (+ next.|init_pre_p1.y| next.|delta_y|))
      (= next.|pre_p2.x| state.|p2.x|)
      (ite
         state.|p.x:init|
         (= next.|p.x| 0)
         (= next.|p.x| next.|pre_p2.x|))
      (= next.|p.x:init| false)
      (= next.|delta.x| next.|delta_x|)
      (= next.|q.x| next.|delta.x|)
      (=
         next.|x_coord|
         (+ next.|p.x| next.|q.x|))
      (= next.|p2.x| next.|x_coord|)
      (= next.|pre_p2.y| state.|p2.y|)
      (ite
         state.|p.y:init|
         (= next.|p.y| 0)
         (= next.|p.y| next.|pre_p2.y|))
      (= next.|p.y:init| false)
      (= next.|delta.y| next.|delta_y|)
      (= next.|q.y| next.|delta.y|)
      (=
         next.|y_coord|
         (+ next.|p.y| next.|q.y|))
      (= next.|p2.y| next.|y_coord|)
      (=
         next.ok1
         (and
            (= next.|p1.x| next.|p2.x|)
            (= next.|p1.y| next.|p2.y|)))
      (=
         next.cex1
         (or
            (not
               (= next.|p1.x| 20))
            (not
               (= next.|p1.y| 13))))
      (= next.|z:1| next.|delta_x|)
      (=
         next.|abs:1|
         (ite
            (<= 0 next.|z:1|)
            next.|z:1|
            (- next.|z:1|)))
      (= next.z next.|delta_y|)
      (=
         next.abs
         (ite
            (<= 0 next.z)
            next.z
            (- next.z)))
      (=
         next.|Leq:1|
         (<= next.|abs:1| 2))
      (=
         next.Leq
         (<= next.abs 2))))

(query
   TS
   (or |gal-initializing| cex1))
(query
   TS
   (or |gal-initializing| ok1))

[Lustre] Validating properties:
[Lustre] Property cex1...
  [Sally][Invalid]
[Trace] 
cex1:
 Step | delta_x | delta_y | |-> 
------+---------+---------+-----
 1    | 2       | 2       |     
 2    | 2       | 2       |     
 3    | 2       | 2       |     
 4    | 2       | 2       |     
 5    | 2       | 2       |     
 6    | 2       | 2       |     
 7    | 2       | 2       |     
 8    | 2       | -1      |     
 9    | 2       | 0       |     
 10   | 2       | 0       |     

[Lustre] Property ok1...
  [Sally][Valid] using 1 past state
[Lustre] Summary:
  [Valid] 1
  [Unknown] 0
  [Invalid] 1
[Lustre] Model status: [Invalid]
